File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailFragment.java
/*
 * Copyright (C) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.android.architecture.blueprints.todoapp.taskdetail;

import android.app.Activity;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.TextView;

import com.example.android.architecture.blueprints.todoapp.R;
import com.example.android.architecture.blueprints.todoapp.addedittask.AddEditTaskActivity;
import com.example.android.architecture.blueprints.todoapp.addedittask.AddEditTaskFragment;
import com.example.android.architecture.blueprints.todoapp.mvibase.MviIntent;
import com.example.android.architecture.blueprints.todoapp.mvibase.MviView;
import com.example.android.architecture.blueprints.todoapp.mvibase.MviViewModel;
import com.example.android.architecture.blueprints.todoapp.mvibase.MviViewState;
import com.example.android.architecture.blueprints.todoapp.util.ToDoViewModelFactory;
import com.jakewharton.rxbinding2.view.RxView;

import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.subjects.PublishSubject;

/**
 * Main UI for the task detail screen.
 */
public class TaskDetailFragment extends Fragment implements MviView<TaskDetailIntent, TaskDetailViewState> {

    @NonNull
    private static final String ARGUMENT_TASK_ID = "TASK_ID";

    @NonNull
    private static final int REQUEST_EDIT_TASK = 1;

    private TextView mDetailTitle;
    private TextView mDetailDescription;
    private CheckBox mDetailCompleteStatus;
    private FloatingActionButton fab;

    TaskDetailViewModel mViewModel;

    // Used to manage the data flow lifecycle and avoid memory leak.
    private CompositeDisposable mDisposables = new CompositeDisposable();
    private PublishSubject<TaskDetailIntent.DeleteTask> mDeleteTaskIntentPublisher = PublishSubject.create();

    public static TaskDetailFragment newInstance(@Nullable String taskId) {
        Bundle arguments = new Bundle();
        arguments.putString(ARGUMENT_TASK_ID, taskId);
        TaskDetailFragment fragment = new TaskDetailFragment();
        fragment.setArguments(arguments);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.taskdetail_frag, container, false);
        setHasOptionsMenu(true);
        mDetailTitle = (TextView) root.findViewById(R.id.task_detail_title);
        mDetailDescription = (TextView) root.findViewById(R.id.task_detail_description);
        mDetailCompleteStatus = (CheckBox) root.findViewById(R.id.task_detail_complete);

        // Set up floating action button
        fab = (FloatingActionButton) getActivity().findViewById(R.id.fab_edit_task);

        return root;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        mViewModel = ViewModelProviders.of(this, ToDoViewModelFactory.getInstance(getContext()))
                .get(TaskDetailViewModel.class);
        mDisposables = new CompositeDisposable();
        bind();
    }

    /**
     * Connect the {@link MviView} with the {@link MviViewModel}
     * We subscribe to the {@link MviViewModel} before passing it the {@link MviView}'s {@link MviIntent}s.
     * If we were to pass {@link MviIntent}s to the {@link MviViewModel} before listening to it,
     * emitted {@link MviViewState}s could be lost
     */
    private void bind() {
        // Subscribe to the ViewModel and call render for every emitted state
        mDisposables.add(mViewModel.states().subscribe(this::render));
        // Pass the UI's intents to the ViewModel
        mViewModel.processIntents(intents());

        // Debounce the FAB clicks to avoid consecutive clicks and navigate to EditTask
        RxView.clicks(fab).debounce(200, TimeUnit.MILLISECONDS)
                .subscribe(view -> showEditTask(getArgumentTaskId()));
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mDisposables.dispose();
    }

    @Override
    public Observable<TaskDetailIntent> intents() {
        return Observable.merge(initialIntent(), checkBoxIntents(), deleteIntent());
    }

    /**
     * The initial Intent the {@link MviView} emit to convey to the {@link MviViewModel}
     * that it is ready to receive data.
     * This initial Intent is also used to pass any parameters the {@link MviViewModel} might need
     * to render the initial {@link MviViewState} (e.g. the task id to load).
     */
    private Observable<TaskDetailIntent.InitialIntent> initialIntent() {
        return Observable.just(TaskDetailIntent.InitialIntent.create(getArgumentTaskId()));
    }

    private Observable<TaskDetailIntent> checkBoxIntents() {
        return RxView.clicks(mDetailCompleteStatus).map(
                activated -> {
                    if (mDetailCompleteStatus.isChecked()) {
                        return TaskDetailIntent.CompleteTaskIntent.create(getArgumentTaskId());
                    } else {
                        return TaskDetailIntent.ActivateTaskIntent.create(getArgumentTaskId());
                    }
                }
        );

    }

    private Observable<TaskDetailIntent.DeleteTask> deleteIntent() {
        return mDeleteTaskIntentPublisher;
    }

    @Nullable
    private String getArgumentTaskId() {
        Bundle args = getArguments();
        if (args == null) return null;
        return args.getString(ARGUMENT_TASK_ID);
    }

    @Override
    public void render(TaskDetailViewState state) {

        setLoadingIndicator(state.loading());

        if (!state.title().isEmpty()) {
            showTitle(state.title());
        } else {
            hideTitle();
        }

        if (!state.description().isEmpty()) {
            showDescription(state.description());
        } else {
            hideDescription();
        }

        showActive(state.active());

        if (state.taskComplete()) {
            showTaskMarkedComplete();
        }

        if (state.taskActivated()) {
            showTaskMarkedActive();
        }

        if (state.taskDeleted()) {
            getActivity().finish();
        }

    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_delete:
                mDeleteTaskIntentPublisher.onNext(TaskDetailIntent.DeleteTask.create(getArgumentTaskId()));
                return true;
        }
        return false;
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.taskdetail_fragment_menu, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_EDIT_TASK) {
            // If the task was edited successfully, go back to the list.
            if (resultCode == Activity.RESULT_OK) {
                getActivity().finish();
                return;
            }
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    public void setLoadingIndicator(boolean active) {
        if (active) {
            mDetailTitle.setText("");
            mDetailDescription.setText(getString(R.string.loading));
        }
    }

    public void hideDescription() {
        mDetailDescription.setVisibility(View.GONE);
    }

    public void hideTitle() {
        mDetailTitle.setVisibility(View.GONE);
    }

    public void showActive(boolean isActive) {
        mDetailCompleteStatus.setChecked(!isActive);
    }

    public void showDescription(@NonNull String description) {
        mDetailDescription.setVisibility(View.VISIBLE);
        mDetailDescription.setText(description);
    }

    private void showEditTask(@NonNull String taskId) {
        Intent intent = new Intent(getContext(), AddEditTaskActivity.class);
        intent.putExtra(AddEditTaskFragment.ARGUMENT_EDIT_TASK_ID, taskId);
        startActivityForResult(intent, REQUEST_EDIT_TASK);
    }

    public void showTaskMarkedComplete() {
        Snackbar.make(getView(), getString(R.string.task_marked_complete), Snackbar.LENGTH_LONG)
                .show();
    }

    public void showTaskMarkedActive() {
        Snackbar.make(getView(), getString(R.string.task_marked_active), Snackbar.LENGTH_LONG)
                .show();
    }

    public void showTitle(@NonNull String title) {
        mDetailTitle.setVisibility(View.VISIBLE);
        mDetailTitle.setText(title);
    }

    public void showMissingTask() {
        mDetailTitle.setText("");
        mDetailDescription.setText(getString(R.string.no_data));
    }

}

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailFragment.kt
/*
 * Copyright (C) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.android.architecture.blueprints.todoapp.taskdetail

import android.app.Activity
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.TextView
import com.example.android.architecture.blueprints.todoapp.R
import com.example.android.architecture.blueprints.todoapp.addedittask.AddEditTaskActivity
import com.example.android.architecture.blueprints.todoapp.addedittask.AddEditTaskFragment
import com.example.android.architecture.blueprints.todoapp.mvibase.MviIntent
import com.example.android.architecture.blueprints.todoapp.mvibase.MviView
import com.example.android.architecture.blueprints.todoapp.mvibase.MviViewModel
import com.example.android.architecture.blueprints.todoapp.mvibase.MviViewState
import com.example.android.architecture.blueprints.todoapp.taskdetail.TaskDetailIntent.ActivateTaskIntent
import com.example.android.architecture.blueprints.todoapp.taskdetail.TaskDetailIntent.CompleteTaskIntent
import com.example.android.architecture.blueprints.todoapp.util.ToDoViewModelFactory
import com.jakewharton.rxbinding2.view.RxView
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.PublishSubject
import java.util.concurrent.TimeUnit
import kotlin.LazyThreadSafetyMode.NONE

/**
 * Main UI for the task detail screen.
 */
class TaskDetailFragment : Fragment(), MviView<TaskDetailIntent, TaskDetailViewState> {
  private lateinit var detailTitle: TextView
  private lateinit var detailDescription: TextView
  private lateinit var detailCompleteStatus: CheckBox
  private lateinit var fab: FloatingActionButton

  private val viewModel: TaskDetailViewModel by lazy(NONE) {
    ViewModelProviders
        .of(this, ToDoViewModelFactory.getInstance(context!!))
        .get(TaskDetailViewModel::class.java)
  }

  // Used to manage the data flow lifecycle and avoid memory leak.
  private var disposables = CompositeDisposable()
  private val deleteTaskIntentPublisher = PublishSubject.create<TaskDetailIntent.DeleteTask>()

  private val argumentTaskId: String
    get() = arguments!!.getString(ARGUMENT_TASK_ID)

  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
      savedInstanceState: Bundle?): View? {
    val root = inflater.inflate(R.layout.taskdetail_frag, container, false)
    setHasOptionsMenu(true)
    detailTitle = root.findViewById<View>(R.id.task_detail_title) as TextView
    detailDescription = root.findViewById<View>(R.id.task_detail_description) as TextView
    detailCompleteStatus = root.findViewById<View>(R.id.task_detail_complete) as CheckBox

    // Set up floating action button
    fab = activity!!.findViewById<View>(R.id.fab_edit_task) as FloatingActionButton

    return root
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)

    bind()
  }

  /**
   * Connect the [MviView] with the [MviViewModel]
   * We subscribe to the [MviViewModel] before passing it the [MviView]'s [MviIntent]s.
   * If we were to pass [MviIntent]s to the [MviViewModel] before listening to it,
   * emitted [MviViewState]s could be lost
   */
  private fun bind() {
    // Subscribe to the ViewModel and call render for every emitted state
    disposables.add(viewModel.states().subscribe(this::render))
    // Pass the UI's intents to the ViewModel
    viewModel.processIntents(intents())

    // Debounce the FAB clicks to avoid consecutive clicks and navigate to EditTask
    RxView.clicks(fab).debounce(200, TimeUnit.MILLISECONDS)
        .subscribe { showEditTask(argumentTaskId) }
  }

  override fun onDestroy() {
    super.onDestroy()
    disposables.dispose()
  }

  override fun intents(): Observable<TaskDetailIntent> {
    return Observable.merge(initialIntent(), checkBoxIntents(), deleteIntent())
  }

  /**
   * The initial Intent the [MviView] emit to convey to the [MviViewModel]
   * that it is ready to receive data.
   * This initial Intent is also used to pass any parameters the [MviViewModel] might need
   * to render the initial [MviViewState] (e.g. the task id to load).
   */
  private fun initialIntent(): Observable<TaskDetailIntent.InitialIntent> {
    return Observable.just(TaskDetailIntent.InitialIntent(argumentTaskId))
  }

  private fun checkBoxIntents(): Observable<TaskDetailIntent> {
    return RxView.clicks(detailCompleteStatus).map {
      if (detailCompleteStatus.isChecked) {
        CompleteTaskIntent(argumentTaskId)
      } else {
        ActivateTaskIntent(argumentTaskId)
      }
    }

  }

  private fun deleteIntent(): Observable<TaskDetailIntent.DeleteTask> {
    return deleteTaskIntentPublisher
  }

  override fun render(state: TaskDetailViewState) {
    setLoadingIndicator(state.loading)

    if (!state.title.isEmpty()) {
      showTitle(state.title)
    } else {
      hideTitle()
    }

    if (!state.description.isEmpty()) {
      showDescription(state.description)
    } else {
      hideDescription()
    }

    showActive(state.active)

    if (state.taskComplete) {
      showTaskMarkedComplete()
    }

    if (state.taskActivated) {
      showTaskMarkedActive()
    }

    if (state.taskDeleted) {
      activity!!.finish()
    }

  }

  override fun onOptionsItemSelected(item: MenuItem?): Boolean {
    when (item!!.itemId) {
      R.id.menu_delete -> {
        deleteTaskIntentPublisher.onNext(TaskDetailIntent.DeleteTask(argumentTaskId))
        return true
      }
    }
    return false
  }

  override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
    inflater!!.inflate(R.menu.taskdetail_fragment_menu, menu)
    super.onCreateOptionsMenu(menu, inflater)
  }

  override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
    if (requestCode == REQUEST_EDIT_TASK) {
      // If the task was edited successfully, go back to the list.
      if (resultCode == Activity.RESULT_OK) {
        activity!!.finish()
        return
      }
    }
    super.onActivityResult(requestCode, resultCode, data)
  }

  fun setLoadingIndicator(active: Boolean) {
    if (active) {
      detailTitle.text = ""
      detailDescription.text = getString(R.string.loading)
    }
  }

  fun hideDescription() {
    detailDescription.visibility = View.GONE
  }

  fun hideTitle() {
    detailTitle.visibility = View.GONE
  }

  fun showActive(isActive: Boolean) {
    detailCompleteStatus.isChecked = !isActive
  }

  fun showDescription(description: String) {
    detailDescription.visibility = View.VISIBLE
    detailDescription.text = description
  }

  private fun showEditTask(taskId: String) {
    val intent = Intent(context, AddEditTaskActivity::class.java)
    intent.putExtra(AddEditTaskFragment.ARGUMENT_EDIT_TASK_ID, taskId)
    startActivityForResult(intent, REQUEST_EDIT_TASK)
  }

  fun showTaskMarkedComplete() {
    Snackbar.make(view!!, getString(R.string.task_marked_complete), Snackbar.LENGTH_LONG)
        .show()
  }

  fun showTaskMarkedActive() {
    Snackbar.make(view!!, getString(R.string.task_marked_active), Snackbar.LENGTH_LONG)
        .show()
  }

  fun showTitle(title: String) {
    detailTitle.visibility = View.VISIBLE
    detailTitle.text = title
  }

  companion object {
    private const val ARGUMENT_TASK_ID = "TASK_ID"
    private const val REQUEST_EDIT_TASK = 1

    operator fun invoke(taskId: String): TaskDetailFragment {
      return TaskDetailFragment().apply {
        arguments = Bundle().apply {
          putString(ARGUMENT_TASK_ID, taskId)
        }
      }
    }
  }
}
